Detectors/TPC/spacecharge/test/testO2TPCPoissonSolver.cxx

```diff
@@ -0,0 +1,240 @@
+    }
+  }
+
+  for (size_t iPhi = 0; iPhi < Nphi; ++iPhi) {
+    const DataT phi = getPhiVertex<DataT, Nz, Nr, Nphi>(iPhi, grid);
+    for (size_t iR = 0; iR < Nr; ++iR) {
+      const DataT radius = getRVertex<DataT, Nz, Nr, Nphi>(iR, grid);
+      const size_t iZ = 0;
+      const DataT z = getZVertex<DataT, Nz, Nr, Nphi>(iZ, grid);
+      potential(iZ, iR, iPhi) = formulas.evalPotential(z, radius, phi);
+    }
+  }
+
+  for (size_t iPhi = 0; iPhi < Nphi; ++iPhi) {
+    const DataT phi = getPhiVertex<DataT, Nz, Nr, Nphi>(iPhi, grid);
+    for (size_t iR = 0; iR < Nr; ++iR) {
+      const DataT radius = getRVertex<DataT, Nz, Nr, Nphi>(iR, grid);
+      const size_t iZ = Nz - 1;
+      const DataT z = getZVertex<DataT, Nz, Nr, Nphi>(iZ, grid);
+      potential(iZ, iR, iPhi) = formulas.evalPotential(z, radius, phi);
+    }
+  }
+}
+
+template <typename DataT, size_t Nz, size_t Nr, size_t Nphi>
+void testAlmostEqualArray(o2::tpc::DataContainer3D<DataT, Nz, Nr, Nphi>& analytical, o2::tpc::DataContainer3D<DataT, Nz, Nr, Nphi>& numerical)
+{
+  for (size_t iPhi = 0; iPhi < Nphi; ++iPhi) {
+    for (size_t iR = 0; iR < Nr; ++iR) {
+      for (size_t iZ = 0; iZ < Nz; ++iZ) {
+        if (std::fabs(analytical(iZ, iR, iPhi)) < ABSTOLERANCE) {
+          BOOST_CHECK_SMALL(numerical(iZ, iR, iPhi) - analytical(iZ, iR, iPhi), ABSTOLERANCE);
+        } else {
+          BOOST_CHECK_CLOSE(numerical(iZ, iR, iPhi), analytical(iZ, iR, iPhi), TOLERANCE);
+        }
+      }
+    }
+  }
+}
+
+template <typename DataT, size_t Nz, size_t Nr, size_t Nphi>
+void testAlmostEqualArray2D(o2::tpc::DataContainer3D<DataT, Nz, Nr, Nphi>& analytical, o2::tpc::DataContainer3D<DataT, Nz, Nr, Nphi>& numerical)
+{
+  for (size_t iPhi = 0; iPhi < Nphi; ++iPhi) {
+    for (size_t iR = 0; iR < Nr; ++iR) {
+      for (size_t iZ = 0; iZ < Nz; ++iZ) {
+        if (std::fabs(analytical(iZ, iR, iPhi)) < ABSTOLERANCE) {
+          BOOST_CHECK_SMALL(numerical(iZ, iR, iPhi) - analytical(iZ, iR, iPhi), ABSTOLERANCE);
+        } else {
+          BOOST_CHECK_CLOSE(numerical(iZ, iR, iPhi), analytical(iZ, iR, iPhi), TOLERANCE2D);
+        }
+      }
+    }
+  }
+}
+
+template <typename DataT, size_t Nz, size_t Nr, size_t Nphi>
+void poissonSolver3D()
+{
+  using GridProp = GridProperties<DataT, Nr, Nz, Nphi>;
+  const o2::tpc::RegularGrid3D<DataT, Nz, Nr, Nphi> grid3D{GridProp::ZMIN, GridProp::RMIN, GridProp::PHIMIN, GridProp::GRIDSPACINGZ, GridProp::GRIDSPACINGR, GridProp::GRIDSPACINGPHI};
+
+  using DataContainer = o2::tpc::DataContainer3D<DataT, Nz, Nr, Nphi>;
+  DataContainer potentialNumerical{};
+  DataContainer potentialAnalytical{};
+  DataContainer charge{};
+
+  const o2::tpc::AnalyticalFields<DataT> analyticalFields;
+  // set the boudnary and charge for numerical poisson solver
+  setChargeDensityFromFormula<DataT, Nz, Nr, Nphi>(analyticalFields, grid3D, charge);
+  setPotentialBoundaryFromFormula<DataT, Nz, Nr, Nphi>(analyticalFields, grid3D, potentialNumerical);
```

Comment:
Is there a reason why you set the boundary potentials of the numerical calculation with the values from the analytical formula? The numerical solution should work fine without this as it should calculate the potential at these boundaries from the charge? 

The function to set only the boundary potentials is supposed to provide to possibility to account for irregularities of the nominal potentials, e.g. charge-up effects at dielectric sector boundaries.

Reply:
> Is there a reason why you set the boundary potentials of the numerical calculation with the values from the analytical formula?

It was like this in the original code. So I always used theses as the boundary conditions.

> The numerical solution should work fine without this as it should calculate the potential at these boundaries from the charge?

I am not sure if boundary conditions are really needed or not for this method. I made a quick comparison by setting the boundary potential and then calculating the potential with the poisonsolver and without setting the boundary potential. The difference is quit big between the resulting potential to the reference potential. See the two plots "analytical potential - numerical potential". The first plot is with the boundary conditions and the setting plot without these boundary conditions for the potential.


![with_boundary_conditions](https://user-images.githubusercontent.com/48915672/97005999-7b425600-153f-11eb-9b76-9ca79ac615af.png)
![no_boundary_conditions](https://user-images.githubusercontent.com/48915672/97006055-8e552600-153f-11eb-80bb-618898edcb98.png)

Reply:
Ok. Now I am thinking about what this means for the case when we use density histograms as input. Maybe at some point we can try to put there the nominal TPC boundary conditions (potential CE, G1T and field cage) already for the Poisson solver instead only adding it (Ez0=400 V/cm) for the calculation of the distortions.